### PR TITLE
Improve mutate function

### LINKED ISSUE #253: Accessing cached data outside hooks.
I think its crucial to access cached data, at least while mutating.

lets say I have 2 endpoints - `api/users` & `api/user/<id>`

say i update `api/user/<id>`, i do not want to invalidate `api/users`, I would just like to mutate it, something like this - `mutate('api/users', [...usersWithoutModifiedUser, modifiedUser])`

the problem is.. I can't get the cached value for `api/users` at the time of the mutation.. since I only have the single user data I'm modifying.

I was thinking mutation api should look something like this

```
mutate('api/users', (cachedData) => {
  ... // using the cached data to build the new cached data.
   return [...usersWithoutModifiedUser, modifiedUser];
})
```

Hope you understand my issue, looking forward hearing from you,
Thank you for the great lib.

### PR DESCRIPTION
**Pull request by sergiodxa**
- If no data call trigger against the key
- If data is a function call it passing current cached value
- Return data or throw error after function
- SWR returns bound mutate function with the `key` already defined

Resolves https://github.com/zeit/swr/issues/155
Resolves #253

### CODE REVIEW

**Comment by jameslouiz** [comment]
This is exactly what I'm looking for. Great work!

**Comment by omerman** [comment]
> This is exactly what I'm looking for. Great work!
> 
> In the mean time, for a bounded Mutate with ability to provide a function with access to the old data you can do this:
> 
> ```
> const useSWRWithBoundMutate = (key, fetcher, options) => {
>   const { data, error } = useSWR(key, fetcher, options);
>   const boundMutate = fn => mutate(key, fn(data));
>   return { data, error, boundMutate };
> };
> ```

It also gives you freedom of accessing the mutate from wherever you want, not only via react's scope( useSWR hook), because in many cases you would want to update just a part of a given response, and it is impossible to perform it outside react's scope with the current api.

**Comment by quietshu** [comment]
Also as I commented in #262, I think we should remove the `shouldRevalidate` trick in https://github.com/zeit/swr/blob/edcfd265d3ad22901b203ecdb25a1fdc92254bf2/src/use-swr.ts#L112-L116 to avoid misunderstanding.

**Comment by sergiodxa** [comment]
@quietshu I updated the README and removed the `shouldRevalidate`, always default to `true` now. Also updated the Optimistic UI example.

File: test/use-swr.test.tsx [hunk 1]
```diff
@@ -847,7 +847,8 @@ describe('useSWR - local mutation', () => {
       // mutate and revalidate
```
Comment (omerman) [inline test/use-swr.test.tsx]: should be just `// mutate` :P

File: README.md [hunk 2]
```diff
@@ -354,6 +357,31 @@ Here’s an example showing the “local mutate - request - update” usage:
 mutate('/api/user', newUser, false)      // use `false` to mutate without revalidation
 mutate('/api/user', updateUser(newUser)) // `updateUser` is a Promise of the request,
                                          // which returns the updated document
```
Comment (quietshu) [inline README.md]: I think there're 3 missing backticks ` ``` ` after this line.

Comment (sergiodxa) [reply README.md]: Done

File: src/use-swr.ts [hunk 3]
```diff
@@ -90,16 +90,30 @@ const broadcastState: broadcastStateInterface = (key, data, error) => {
...
-const mutate: mutateInterface = async (_key, _data, shouldRevalidate) => {
+const mutate: mutateInterface = async (
+  _key,
+  _data,
+  shouldRevalidate = true
+) => {
   const [key] = getKeyArgs(_key)
   if (!key) return
 
+  // if there is no new data data, call revalidate against the key
```
Comment (quietshu) [inline src/use-swr.ts]:
```suggestion
  // if there is no new data, call revalidate against the key
```

File: src/use-swr.ts [hunk 4]
```diff
@@ -90,16 +90,30 @@ const broadcastState: broadcastStateInterface = (key, data, error) => {
...
+  if (!_data) return trigger(_key)
```
Comment (quietshu) [inline src/use-swr.ts]:
```suggestion
  if (typeof _data === 'undefined') return trigger(_key, shouldRevalidate)
```

File: test/use-swr.test.tsx [hunk 5]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
     })
     await act(() => new Promise(res => setTimeout(res, 110)))
     expect(container.textContent).toMatchInlineSnapshot(`"data: 999"`)
   })
 
+  it('should trigger on mutation without data', async () => {
+    let value = 0
+
+    function Page() {
+      const { data } = useSWR('dynamic-12', () => value++, {
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
      const { data } = useSWR('dynamic-14', () => value++, {
```

File: test/use-swr.test.tsx [hunk 6]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+    }
+    const { container } = render(<Page />)
+
+    // hydration
+    expect(container.firstChild.textContent).toMatchInlineSnapshot(`"data: "`)
+    await waitForDomChange({ container }) // mount
+    expect(container.firstChild.textContent).toMatchInlineSnapshot(`"data: 0"`)
+    await act(() => {
+      // trigger revalidation
+      mutate('dynamic-12')
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
      mutate('dynamic-14')
```

File: test/use-swr.test.tsx [hunk 7]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+      return new Promise(res => setTimeout(res, 1))
+    })
+    expect(container.firstChild.textContent).toMatchInlineSnapshot(`"data: 1"`)
+  })
+
+  it('should call function as data passing current cached value', async () => {
+    // prefill cache with data
+    cacheSet('dynamic-13', 'cached data')
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
    cacheSet('dynamic-15', 'cached data')
```

File: test/use-swr.test.tsx [hunk 8]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+    const callback = jest.fn()
+    await mutate('dynamic-13', callback)
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
    await mutate('dynamic-15', callback)
```

File: test/use-swr.test.tsx [hunk 9]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+    expect(callback).toHaveBeenCalledWith('cached data')
+  })
+
+  it('should return results of the mutation', async () => {
+    // returns the data if promise resolved
+    expect(mutate('dynamic-14', Promise.resolve('data'))).resolves.toBe('data')
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
    expect(mutate('dynamic-16', Promise.resolve('data'))).resolves.toBe('data')
```

File: test/use-swr.test.tsx [hunk 10]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+
+    // throw the error if promise rejected
+    expect(
+      mutate('dynamic-14', Promise.reject(new Error('error')))
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
      mutate('dynamic-16', Promise.reject(new Error('error')))
```

File: test/use-swr.test.tsx [hunk 11]
```diff
@@ -846,13 +847,82 @@ describe('useSWR - local mutation', () => {
...
+    ).rejects.toBeInstanceOf(Error)
+  })
+
+  it('should get bound mutate from useSWR', async () => {
+    function Page() {
+      // eslint-disable-next-line no-shadow
+      const { data, mutate: boundMutate } = useSWR(
+        'dynamic-15',
```
Comment (quietshu) [inline test/use-swr.test.tsx]:
```suggestion
        'dynamic-17',
```

**Comment by quietshu** [comment]
Thank you @sergiodxa! 

Close #262, #136.

**Pull request merged**